src/libraries/System.Private.CoreLib/src/System/SearchValues/IndexOfAnyAsciiSearcher.cs

```diff
@@ -81,14 +94,87 @@ internal static unsafe void ComputeAsciiState<T>(ReadOnlySpan<T> values, out Asc
+
+        public static bool CanUseUniqueLowNibbleSearch<T>(ReadOnlySpan<T> values, int minInclusive, int maxInclusive)
+            where T : struct, IUnsignedNumber<T>
+        {
+            Debug.Assert(typeof(T) == typeof(byte) || typeof(T) == typeof(char));
 
-                int highNibble = value >> 4;
-                int lowNibble = value & 0xF;
+            if (!IsVectorizationSupported || values.Length > 16)
+            {
+                return false;
+            }
 
-                bitmapLocal[(uint)lowNibble] |= (byte)(1 << highNibble);
+            if (Ssse3.IsSupported && maxInclusive > 127)
+            {
+                // We could support [1, 254] if we did the "& 0xF" before calling into Shuffle in IndexOfAnyLookupUniqueLowNibbleCore.
+                // We currently optimize for the common case of ASCII characters instead, saving an instruction there.
+                return false;
             }
 
-            state = new AsciiState(bitmapSpace, lookupLocal);
+            if (typeof(T) == typeof(char))
+            {
+                if (maxInclusive >= byte.MaxValue)
+                {
+                    // When packing UTF-16 characters into bytes, values may saturate to 255 (false positives), hence ">=" instead of ">".
+                    return false;
+                }
+
+                if ((Ssse3.IsSupported || PackedSimd.IsSupported) && minInclusive == 0)
+                {
+                    // When packing UTF-16 characters into bytes, values may saturate to 0 (false positives).
+                    // We could also support 0 if we flowed Ssse3AndWasmHandleZeroInNeedle through.
+                    return false;
+                }
+            }
+
+            // We assume there are no duplicates to simplify the logic (if there are any, they just won't use this searching approach).
+            int seenNibbles = 0;
+
+            foreach (T tValue in values)
+            {
+                int bit = 1 << (int.CreateChecked(tValue) & 0xF);
+
+                if ((seenNibbles & bit) != 0)
+                {
+                    // We already saw a value with the same low nibble.
+                    return false;
+                }
+
+                seenNibbles |= bit;
+            }
+
+            return true;
+        }
+
+        public static void ComputeUniqueLowNibbleState<T>(ReadOnlySpan<T> values, out UniqueLowNibbleState state)
+            where T : struct, IUnsignedNumber<T>
+        {
+            Debug.Assert(typeof(T) == typeof(byte) || typeof(T) == typeof(char));
+
+            Vector128<byte> valuesByLowNibble = default;
+            BitVector256 lookup = default;
+
+            foreach (T tValue in values)
+            {
+                byte value = byte.CreateTruncating(tValue);
+                lookup.Set(value);
+                valuesByLowNibble.SetElementUnsafe(value & 0xF, value);
+            }
+
+            if (valuesByLowNibble.GetElement(0) == 0 && !lookup.Contains(0))
```

Comment:
This assumes the values are sorted?

Reply:
No, `valuesByLowNibble` will have the same value regardless of the input order.